### ISSUE TITLE
Duplicate calls to method

### ISSUE THREAD

**Issue by nixpulvis** [egonSchiele/contracts.ruby#130]
This is a pretty nasty bug I bumped into today. These kinds of things scare me *a lot*.

```ruby
require 'contracts'

class Foo
  include Contracts

  class << self
    Contract None => Num
    def self.bar
      123
    end

    Contract None => Any
    def foo
      puts "123"
    end
  end
end

Foo.foo
```

Running this puts

```
123
123
```

which is not correct.

I'm running the gem package 0.8.

**Comment by egonSchiele** [comment]
Wow, that's super weird. I can reproduce. The issue goes away if I change `self.bar` to `bar`:

    class Foo
      include Contracts

      class << self
        Contract None => Num
        def bar
          123
        end

        Contract None => Any
        def foo
          puts "123"
        end
      end
    end

    Foo.foo

Still, wonder what is causing this.

**Comment by nixpulvis** [comment]
Yea, it's not actually causing me pain because I stumble upon this by refactoring some code from the `self.<method>` style to the `class << self` style. However it's still worth looking into.

**Comment by egonSchiele** [comment]
Agreed, contracts should never change behavior.

**Comment by waterlink** [comment]
I don't really understand, how `Foo` gets `.bar` method in first example.. What happens if you remove contract? Will it fail with undefined method?

**Comment by egonSchiele** [comment]
If you remove the contract on `self.bar`, it just runs `foo` once. If you remove the contract on `foo`, it throws this error:

    /Users/adit/.rvm/gems/ruby-2.1.2/gems/contracts-0.8/lib/contracts.rb:76:in `block in <class:Contract>': Contract violation for return value: (ContractError)
            Expected: Num,
            Actual: nil
            Value guarded in: Foo::foo
            With Contract: None => Num
            At: test.rb:12

So it's using the contract for `bar` on `foo`. My guess is it's something to do with method overloading.

**Comment by egonSchiele** [comment]
The issue is, we create the first decorator, `Contract None => Num`. Then we add the function `self.bar`, but that doesn't trigger a `method_added`. Then we add another contract, and another function, `foo`. `foo` gets assigned *both* contracts. That's the first mistake.

Now we treat `foo` as if there are two versions of the function (multi-dispatch). Here are the types for both the versions:

    Contract None => Num

and

    Contract None => Any

obviously this is not valid multi-dispatch syntax, because the params are the same and the *return values* are different. So `foo` gets executed once. The return value doesn't match `Num`, so we assume that was the wrong version to call. Then it gets executed AGAIN with the second contract. This one matches.

Lots of things going wrong here:

1. `self.bar` should trigger a `method_added`, why didn't it happen?
2. we shouldn't assign the contract for `bar` to `foo` incorrectly.
3. if two versions of a function have the same params, that's not multi-dispatch. We should raise an error or something.
4. Multi-dispatch shouldn't take the return value into account at all. Once we have started running a function, it is too late to `rescue` it and say "this is not the version we meant to run". That should happen only when checking the contract on the params.

Thanks for filing this @nixpulvis , you caught a lot of issues at once.

**Comment by nixpulvis** [comment]
Ironically, the fact that this broke the way it did helped me catch a bug i wrote in my program, but yea glad to help :D

**Comment by waterlink** [comment]
Actually, it haven't broke. It was like that from the beginning. It was
failing through to next overloads when result contract was violated. Bug or
feature ? Probably nasty bug, or dangerous feature :)

--
Best Regards,
Alexey Fedorov,
Sr Ruby Developer, Clojure Developer, Rust Developer,
+49 15757 486 476

2015-04-13 14:59 GMT+02:00 Nathan Lilienthal <notifications@github.com>:

**Comment by egonSchiele** [comment]
Ok, I've got 3 and 4 fixed in branch adit/bug_130. Those were the easy ones. Still not sure why 1 and 2 are breaking.

**Comment by egonSchiele** [comment]
Ok, this code should now raise an error with an error like:

```
Oops, it looks like method 'foo' has multiple contracts:
{Contracts::Num=>Contracts::Num}
{Contracts::None=>Contracts::Any}
```

That's at least better than before. I also pushed a fix so methods won't get more than once (the most alarming part of this bug for me).

I still haven't figured out why `common_method_added` doesn't get called with `self.bar`. All that meta-programming has become tough for me to follow. If either of you guys are interested in taking a look, please do :)

**Comment by nixpulvis** [comment]
>  won't get more than once

Assume you mean "won't get **called** more than once".

**Comment by nixpulvis** [comment]
Nice to have this now fail with an error at least. Still unexpected behavior :(

**Comment by egonSchiele** [comment]
Yup :-/ Just started a new job so I won't have time to work on this for a while. But at least its not calling a method twice anymore...that was the most alarming part for me.

**Comment by nixpulvis** [comment]
Yea, and congrats on the new job :fireworks:

**Comment by egonSchiele** [comment]
Thanks! :+1:

**Comment by nixpulvis** [comment]
This looks fixed to me, can anyone confirm?

**Comment by egonSchiele** [comment]
Confirmed, git bisect says it got fixed in eed57a9354dc5fd7104e5ce4427c41372328037d. Closing.
